core/src/main/java/org/apache/iceberg/io/WriterFactory.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.deletes.EqualityDeleteWriter;
+import org.apache.iceberg.deletes.PositionDeleteWriter;
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+
+/**
+ * A factory for creating data and delete writers.
+ */
+public interface WriterFactory<T> {
+
+  /**
+   * Creates a new {@link DataWriter}.
+   *
+   * @param file the output file
+   * @param spec the partition spec written data belongs to
+   * @param partition the partition written data belongs to or null if the spec is unpartitioned
+   * @return the constructed data writer
+   */
+  DataWriter<T> newDataWriter(EncryptedOutputFile file, PartitionSpec spec, StructLike partition);
+
+  /**
+   * Creates a new {@link EqualityDeleteWriter}.
+   *
+   * @param file the output file
+   * @param spec the partition spec written deletes belong to
+   * @param partition the partition written deletes belong to or null if the spec is unpartitioned
```

Comment:
Maybe better if we have an extra method for nonpartitioned table? Or using `PartitionSpec.unpartitioned()`?
Just some random ideas, since I am not a big fan of null attributes.

Reply:
Ideally, we would have an extra method without the spec and partition. However, we did not reserve any particular spec ID as unpartitioned. That's why we still need a correct spec even for unpartitioned writes as we need to store a correct spec ID in the metadata. We could offer a method without the partition struct but I am not sure how valuable that is.

Reply:
Allowing to set the partition data to be null helps to abstract an uniformed [file-size-rolling data/delete writers](https://github.com/apache/iceberg/blob/63392d95848b21b32e6619f500e87c5342824f08/core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java#L302) in the  BaseTaskWriter,  I mean in this way we don't have to design different writers for unpartitioned table and partitioned tables, that really reduces lots of writer classes.  I think we'd better to follow the existing approach.

Reply:
I think @openinx's point is a good reason to keep the existing behavior and allow nulls.